slynk/slynk.lisp

```diff
@@ -479,8 +479,11 @@ corresponding values in the CDR of VALUE."
 
 ;;; Channels
 
-(defmacro channels () `(connection-channels *emacs-connection*))
-(defmacro channel-counter () `(connection-channel-counter *emacs-connection*))
+(defun channels () (connection-channels *emacs-connection*))
+(defun (setf channels) (value) (setf (connection-channels *emacs-connection*) value))
+
+(defun channel-counter () (connection-channel-counter *emacs-connection*))
+(defun (setf channels) (value) (setf (connection-channel-counter *emacs-connection*) value))
```

Comment:
There's a bug here. And why do you need this again?  I don't understand...

Reply:
Sorry about that. Fixed.

As per why - I had code like
```lisp
(let ((slynk (find-package '#:slynk)))
  (when slynk
    (funcall (find-symbol (string '#:channels) slynk))))
```
and was very surprised to find `channels` be defined as a function.

This is for implementing functionality similar to https://github.com/cbaggers/livesupport so that slynk detection is done at runtime rather than via reader macros which end up badly in situations where you compile the file without slynk and load it with it, or vice versa

Reply:
Then just change it to `connection-channels`, right?

Reply:
I could do that yes, I still think that using a macro where it's unnecessary is bad style, but if you disagree with this or rather not the delta, that's fine too.
I just figured I issue a relatively benign PR

Reply:
For all their disadvantages (such as not being functions), macros provide terseness  and even a bit of runtime efficiency (though I'd be surprised if it mattered in this case). So I'll use them in "private" contexts.  The thing that's surprising me here is that you're hooking into `swank.lisp`'s implementation details.

Can you summarily explain what you want to do? Perhaps an interface can be opened and a symbol exported. One thing that this failures shows is that you're probably missing the right context for the `*emacs-connection*` special.

Reply:
Sure! I'd actually be interested to see how you'd tackle the problem since while the ways it's done right now seem to work, they also seem to break in some weird situations which tells me it might not be 100% stable.

So this is just doing the same as what the [livesupport system I linked before](https://github.com/cbaggers/livesupport) does.

In certain applications, like games, the software is set up so that you have an "update loop" that is spinning on a clock. Livesupport is attempting to add support for being able to have a slime/sly repl which executes within the context of that loop.
The reasons you'd want this (as opposed to executing commands via a worker thread) are:
1. Many objects are thread-sensitive where you can only do meaningful ops on them on a specific thread
2. Even if they are not, 'touching' them from a worker scheduled at an unpredictable time could cause very bad things to happen like accessing objects that have been deleted etc.
3. Theoretically you'd be able to support a repl in a single-threaded implementation by executing requests explicitly inside the update loop

Reply:
> Many objects are thread-sensitive where you can only do meaningful ops on them on a specific thread

Slynk (and Swank) have specific interfaces for interrupting a thread at certain times and running code in specific situations.  These interfaces aren't public, though.

> Theoretically you'd be able to support a repl in a single-threaded implementation by executing requests explicitly inside the update loop

Not only theoretically, in practice that works too.  I use sly from my macbook with an old SBCL which doesn't support multithreading and it works OK.  It uses the so-called `:fd-handler` communication style, which is basically a version of POSIX's `select(2)`. The communication style is controlled via a variable that the user can override: https://joaotavora.github.io/sly/#Communication-style.  If you do that, SLY will use a single thread (and a SIGINT handler).  Which doesn't mean you can't launch more threads manually.

However, none of this seems to be exactly what you want.  

It seems you want something like.

```lisp
(defun game ()
(loop initially (slynk-mrepl::send-prompt slynk::*channel*)
               do (sleep 0.1)
                  (print 'doing-some-game-work)
                  (slynk::process-requests t)))
```

This needs a change to `slynk.lisp`

```diff
diff --git a/slynk/slynk.lisp b/slynk/slynk.lisp
index 5522a74c..97d0831e 100644
--- a/slynk/slynk.lisp
+++ b/slynk/slynk.lisp
@@ -1089,6 +1089,8 @@ The processing is done in the extent of the toplevel restart."
               (with-top-level-restart (connection (go start))
                 (process-requests timeout)))))))
 
+(defvar *channel*)
+
 (defun process-requests (timeout)
   "Read and process requests from Emacs.
 TIMEOUT has the same meaning as in WAIT-FOR-EVENT."
@@ -1101,8 +1103,8 @@ TIMEOUT has the same meaning as in WAIT-FOR-EVENT."
         (when timed-out-p (return))
         (destructure-case event
           ((:emacs-rex &rest args) (apply #'eval-for-emacs args))
-          ((:emacs-channel-send channel (selector &rest args))
-           (channel-send channel selector args)))))))
+          ((:emacs-channel-send *channel* (selector &rest args))
+           (channel-send *channel* selector args)))))))
 
 (defun spawn-channel-thread (connection channel)
   "Spawn a listener thread for CONNECTION and CHANNEL.
```

And, of course, `send-prompt` and `process-requests` have to be exported (maybe `*channel*` doesn't need to).  Anyway, try it out and tell me if it's something like this.

Reply:
> Not only theoretically, in practice that works too. 
Yep! I used to use slime on windows back before SBCL threads were stable on that platform. The 'theoretical' part here is more about actually incorporating & expecting this behaviour in a game loop. It actually makes it -easier- to not have threads there because you have 0 chance of stomping on some state from a C-c C-c or C-c C-e

> It seems you want something like.
Yes, exactly.

Unfortunately I'm not fully understanding what the change to `*channel*` here would be doing.
In the past I've been able to have this functionality by calling `slynk::handle-requests` and passing it `slynk::*emacs-connection*` in the update loop.
But maybe that has some side-effects I'm unaware of

Reply:
> Unfortunately I'm not fully understanding what the change to *channel* here would be doing.

It's just an implementation detail that allows to to call `slynk-mrepl::send-prompt` with the proper argument.  Without it you need some very strange antics to get the `slynk-mrepl::mrepl` object.  

Anyway, never mind the implementation details.  I'm proposing to export `send-prompt` and `process-requests` so you can do a loop like that.  Is that enough for you or do you need more?

> But maybe that has some side-effects I'm unaware of

It's basically the same, but requires access to another special variable.  And misfetching that can cause you to shoot yourself in the foot.

Reply:
Ohhh I see now. So `*channel*` would be accessible from slynk while in the REPL. Got it!

That should work!

While on the subject - I noticed this behaviour when entering the debugger when using `process-requests`:
```lisp
CL-USER> (game) ; this has the call to the game loop
CL-USER> x
; Debugger entered on #<UNBOUND-VARIABLE X {1003B39313}>
[1] CL-USER> 
; Debugger entered on #<UNBOUND-VARIABLE X {1003B39313}>
[2] CL-USER> 
; Evaluation aborted on #<UNBOUND-VARIABLE X {1003B39313}>
[1] CL-USER> "still have debugger indicator on the left"
"still have debugger indicator on the left"
[1] CL-USER> 
```

This is using this definition of `game`:
```lisp
(defun game ()
  (loop initially (slynk-mrepl::send-prompt slynk::*channel*)
        do (sleep 0.1)
           (with-simple-restart (continue "continue")
             (slynk::process-requests t))))
```
Thoughts?